### PR TITLE
[DebugInfo][JumpThreading] Fix missing debug location updates

### LINKED ISSUE #91574: [DebugInfo][JumpThreading] Missing debug location updates
Four missing debug location updates for newly created instructions.

[JumpThreading-L1282](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L1282): The newly created CastInst, which is inserted before the replaced old LoadInst `LoadI`.
```C++
if (AvailableVal->getType() != LoadI->getType())
      AvailableVal = CastInst::CreateBitOrPointerCast(
          AvailableVal, LoadI->getType(), "", LoadI->getIterator());
LoadI->replaceAllUsesWith(AvailableVal);
LoadI->eraseFromParent();
```

[JumpThreading-L2983](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L2983): The newly created PHINode `NewPN`, which is inserted befor the replaced old SelectInst `SI`.
```cpp
PHINode *NewPN = PHINode::Create(SI->getType(), 2, "", SI->getIterator());
NewPN->addIncoming(SI->getTrueValue(), Term->getParent());
NewPN->addIncoming(SI->getFalseValue(), BB);
SI->replaceAllUsesWith(NewPN);
SI->eraseFromParent();
```

[JumpThreading-L3120](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L3120): Newly created PHINode `NewPN`, which is inserted at the beginning of the parent block of the replaced Instruction `Inst`:
```cpp
BasicBlock::iterator InsertionPoint = BB->getFirstInsertionPt();
assert(InsertionPoint != BB->end() && "Empty block?");
 // Substitute with Phis & remove.
for (auto *Inst : reverse(ToRemove)) {
if (!Inst->use_empty()) {
  PHINode *NewPN = PHINode::Create(Inst->getType(), 2);
  NewPN->addIncoming(UnguardedMapping[Inst], UnguardedBlock);
  NewPN->addIncoming(GuardedMapping[Inst], GuardedBlock);
  NewPN->insertBefore(InsertionPoint);
  Inst->replaceAllUsesWith(NewPN);
}
  Inst->dropDbgRecords();
  Inst->eraseFromParent();
}
```

I will give a patch for them later.

### PR DESCRIPTION
**Pull request by Apochens**
Fix #91574.
* [JumpThreading-L2983](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L2983) is tested by `preserving-debugloc-fold-select.ll`.
* [JumpThreading-L3120](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L3120) is tested by `guard-split-debuginfo.ll`.

@SLTozer Could you please help in assigning reviewers for this PR. Thanks! (Now I may not continously work on this project, so I think requesting the contributor permission could be inappropriate. I apologize for directly pinging you again and again.)

It's difficult to construct a suitable test case for [JumpThreading-L1282](https://github.com/llvm/llvm-project/blob/b52fa9461ab73eaf2d04f32c806a1715b2595830/llvm/lib/Transforms/Scalar/JumpThreading.cpp#L1282). It seems that the transformation happens rarely. Probably experienced reviewers can give some suggestions.

### CODE REVIEW

File: llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll [hunk 1]
```diff
@@ -7,6 +7,9 @@
 ; parent blocks. And that ino jump-threading, the old dbg.value gets
 ; deleted.
 
+; Test debug location preserving in function threadGuard, which replaces
+; instructions with PHINodes at the end of the function. [[TMP1]]
```
Comment (OCHyams) [inline llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll:11]: nit: "at the end of the function" are you referring to the code threadGuard? This feels like a comment that could go stale quite easily. I also find the first part a little hard to parse - perhaps the comment could be replaced with something along the lines of "Test that jump threading's `threadGuard` propagates the debug location to the `phi` from the instruction it replaces (`%retVal`)."?

also nit: `[[TMP1]]`?

My comment in the next test about `utils/update_test_checks.py` applies to this one too.

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp [hunk 2]
```diff
@@ -1278,9 +1278,12 @@ bool JumpThreadingPass::simplifyPartiallyRedundantLoad(LoadInst *LoadI) {
     // only happen in dead loops.
     if (AvailableVal == LoadI)
       AvailableVal = PoisonValue::get(LoadI->getType());
-    if (AvailableVal->getType() != LoadI->getType())
+    if (AvailableVal->getType() != LoadI->getType()) {
       AvailableVal = CastInst::CreateBitOrPointerCast(
           AvailableVal, LoadI->getType(), "", LoadI->getIterator());
+      if (Instruction *CI = dyn_cast<Instruction>(AvailableVal))
+        CI->setDebugLoc(LoadI->getDebugLoc());
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp:1285]: re: generating a test for this part, if you stick an `assert(false)` in here and run the build target `check-llvm` do we get any test failures?

**Review by OCHyams — COMMENTED** [review]
Thanks for another helpful fix! Code changes LGTM but I've got a couple of questions about the tests in-line.

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp:1285]: Yes. For each case like this, I would first search for existing tests by sticking an `assert(false)`. If any test is found, I modify that test to obtain a test  for the patched site. For this site, I have not found any existing test, which can trigger it.

File: llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll [hunk 3]
```diff
@@ -0,0 +1,98 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
```
[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: The checks in this test are generated by using update_test_checkes.py, and then I remove some redundant debug metadata checks and keep the `DBG29`. So, if someone regenerate the checks, the crucial checks would still be there.

For those instruction checks, I think we can only keep those of `.exit.thread4`.

[the remark this answers is not in the document]
Comment (pogo59) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: You should not modify or remove checks that were automatically generated. That's the point of automatic generation.

The automated scripts generate far more checks than are actually necessary, but removing some doesn't actually help anything.

[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: > You should not modify or remove checks that were automatically generated. That's the point of automatic generation.

So, should I keep all the automatically generated checks? How about those redundant checks that are irrelevant to the fix in this case?

File: llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll [hunk 4]
```diff
@@ -33,8 +36,9 @@ define i32 @branch_implies_guard(i32 %a) !dbg !7 {
 ; CHECK-NEXT:    br label [[MERGE]]
 ; CHECK:       Merge:
 ; CHECK-NEXT:    [[RETPHI:%.*]] = phi i32 [ [[V1]], [[T1_SPLIT]] ], [ [[V2]], [[F1_SPLIT]] ]
-; CHECK-NEXT:    [[TMP1:%.*]] = phi i32 [ [[RETVAL3]], [[T1_SPLIT]] ], [ [[RETVAL1]], [[F1_SPLIT]] ]
+; CHECK-NEXT:    [[TMP1:%.*]] = phi i32 [ [[RETVAL3]], [[T1_SPLIT]] ], [ [[RETVAL1]], [[F1_SPLIT]] ], !dbg [[DBG12]]
 ; CHECK-NEXT:    ret i32 [[TMP1]], !dbg [[DBG12]]
+; CHECK: [[DBG12]] = !DILocation(line: 13,
```
Comment (OCHyams) [inline llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll:41]: The same concern mentioned below about auto-generated checks applies to this test. The problem is that this test is marked as auto generated so someone might re-generate the test in the future. In this case this would result in a loss of test coverage because the regenerated test would not include your `CHECK: [[DBG12]] = ...` directive.

IMO there are a couple of options to choose from:
1.  Use this file but check metadata too by using the update_test_checks.py argument `--check-globals` 
2. Duplicate this test file, then in the new file remove the "auto-generated" line at the top and the CHECK directives we don't need, and add your new CHECK(s).
3. Use this file but remove the "auto-generated" line at the top.

Those options are more or less in my order of preference, if option 1 doesn't add too much noise to the test - could you try that out?

File: llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll (hunk 3 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: I think removing the line `; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4` here should be enough for this one (discussion continues up in the other test though).

Otherwise the latest change to this one looks good, thanks.

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: That's unfortunate that we have a hole in test coverage here but certainly not unprecedented. Another option could be to use the clang built with that assert to build other codebases, e.g., self-host (build clang again), the llvm-test-suite, or any other codebase you might have access to.

Alternatively, you could try to widen the search, e.g., moving the assert further into common code. If it's inside an `if` at the moment, move it outside the if (or further). When that gives you a test you may be able to modify the IR from that test by hand in order to hit the desired code path.

If you have a go at this and it's not fruitful let me know and I'll have a go at it too.

[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: Ok, I will try other codebases. (Actually, I have tried to remove the assertion out of the `then` branch and found [a test file](https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/JumpThreading/thread-loads.ll) (test8). But I cannot modify it to obtain the desired one.)

File: llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll (hunk 4 above)
Comment (Apochens) [reply llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll:41]: I tried option 1, and checks for all the metadata are automatically generated, adding 17 new checks/lines. Would that be alright?

Comment (OCHyams) [reply llvm/test/Transforms/JumpThreading/guard-split-debuginfo.ll:41]: Thanks, yeah I think that's okay for this one then.

File: llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll (hunk 3 above)
[the remark this answers is not in the document]
Comment (pogo59) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: Right. If the test has a comment saying the tests were auto-generated, that comment needs to tell the truth (no manual modifications of the checks). It's likely the generated CHECKs will do more than you would in a manual test, but that's okay. People using the scripts expect that.

[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/test/Transforms/JumpThreading/preserving-debugloc-fold-select.ll:1]: Oh, I understand now! Thanks for pointing out this!

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: @OCHyams  I have tried `check-llvm`, `check-all` and `check-llvm-unit`, resulting no test triggering the assertion (even if I move the assertion out of the `then` branch).

[the remark this answers is not in the document]
Comment (SLTozer) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: The llvm-test-suite that @OCHyams is referencing isn't the test suite within LLVM (the confusion is understandable of course!), but the [LLVM test suite repo](https://github.com/llvm/llvm-test-suite), which contains a variety of real program selected to test LLVM; it contains a decent spread of programs (so likely to stimulate more code in LLVM) and is faster to build than a stage 2 clang build; there are instructions on how to use it [here](https://llvm.org/docs/TestSuiteGuide.html).

[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: Thanks for trying that out. I managed to stimulate this codepath by modifying the test you linked above - I changed `%a`'s type to float (then ran debugify with the `--debugify-level=locations` option). I haven't added the RUN/CHECK lines but this should do it:
```
target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128"
target triple = "i386-apple-darwin7"

declare void @f1(...)

define void @test8(ptr %0, ptr %1, ptr %2) !dbg !5 {
  %a = load float, ptr %0, align 4, !dbg !8
  %b = load i32, ptr %0, align 4, !dbg !9
  store float %a, ptr %1, align 4, !dbg !10
  %c = icmp eq i32 %b, 8, !dbg !11
  br i1 %c, label %ret1, label %ret2, !dbg !12

ret1:                                             ; preds = %3
  ret void, !dbg !13

ret2:                                             ; preds = %3
  %xxx = tail call i32 (...) @f1() #0, !dbg !14
  ret void, !dbg !15
}

!llvm.dbg.cu = !{!0}
!llvm.debugify = !{!2, !3}
!llvm.module.flags = !{!4}

!0 = distinct !DICompileUnit(language: DW_LANG_C, file: !1, producer: "debugify", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug)
!1 = !DIFile(filename: "test.ll", directory: "/")
!2 = !{i32 8}
!3 = !{i32 0}
!4 = !{i32 2, !"Debug Info Version", i32 3}
!5 = distinct !DISubprogram(name: "test8", linkageName: "test8", scope: null, file: !1, line: 1, type: !6, scopeLine: 1, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0)
!6 = !DISubroutineType(types: !7)
!7 = !{}
!8 = !DILocation(line: 1, column: 1, scope: !5)
!9 = !DILocation(line: 2, column: 1, scope: !5)
!10 = !DILocation(line: 3, column: 1, scope: !5)
!11 = !DILocation(line: 4, column: 1, scope: !5)
!12 = !DILocation(line: 5, column: 1, scope: !5)
!13 = !DILocation(line: 6, column: 1, scope: !5)
!14 = !DILocation(line: 7, column: 1, scope: !5)
!15 = !DILocation(line: 8, column: 1, scope: !5)
```

[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: > The llvm-test-suite that @OCHyams is referencing isn't the test suite within LLVM

Ah, I had intended to link to it and forgot, thanks for clarifying that.

[the remark this answers is not in the document]
Comment (Apochens) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp]: The test works! I have already add checks into it and included it into this PR! Thanks!

**Review by OCHyams — APPROVED** [review]
LGTM, thanks again

**Review by jryans — APPROVED** [review]
Looks good to me as well! 😄 

As before, I'll aim to merge this sometime between tomorrow (Sat) and the next working day (Mon) to allow for any further reviews.
